sys/net/gnrc/link_layer/lwmac/lwmac.c

```diff
@@ -0,0 +1,745 @@
+        break;
+
+    /* Receiving incoming data */
+    case RECEIVING:
+        rtt_handler(LWMAC_EVENT_RTT_PAUSE); /* No duty cycling while TXing */
+        _set_netdev_state(&lwmac, NETOPT_STATE_IDLE);  /* Power up netdev */
+        break;
+
+    case STOPPED:
+        _set_netdev_state(&lwmac, NETOPT_STATE_OFF);
+        break;
+
+    /*********************** Control states ***********************************/
+    case START:
+        rtt_handler(LWMAC_EVENT_RTT_START);
+        lwmac_set_state(LISTENING);
+        break;
+
+    case STOP:
+        rtt_handler(LWMAC_EVENT_RTT_STOP);
+        lwmac_set_state(STOPPED);
+        break;
+
+    case RESET:
+        LOG_WARNING("Reset not yet implemented\n");
+        lwmac_set_state(STOP);
+        lwmac_set_state(START);
+        break;
+
+    /**************************************************************************/
+    default:
+        LOG_DEBUG("No actions for entering state %u\n", newstate);
+        return;
+    }
+
+    lwmac_schedule_update();
+}
+
+/******************************************************************************/
+
+/* Main state machine. Call whenever something happens */
+// TODO: Don't use global variables
+bool lwmac_update(void)
+{
+    lwmac.needs_rescheduling = false;
+
+    switch(lwmac.state)
+    {
+    case SLEEPING:
+
+        /* If a packet is scheduled, no other (possible earlier) packet can be
+         * sent before the first one is handled, even no broadcast
+         */
+        if( !lwmac_timeout_is_running(&lwmac, TIMEOUT_WAIT_FOR_DEST_WAKEUP) ) {
+
+            /* Check if there are broadcasts to send and transmit immediately */
+            if(packet_queue_length(&(lwmac.tx.neighbours[0].queue)) > 0) {
+                lwmac.tx.current_neighbour = &(lwmac.tx.neighbours[0]);
+                lwmac_set_state(TRANSMITTING);
+                break;
+            }
+
+            lwmac_tx_neighbour_t* neighbour = _next_tx_neighbour(&lwmac);
+
+            if(neighbour != NULL) {
+                /* Offset in microseconds when the earliest (phase) destination
+                 * node wakes up that we have packets for. */
+                int time_until_tx = RTT_TICKS_TO_US(_ticks_until_phase(neighbour->phase));
+
+                /* If there's not enough time to prepare a WR to catch the phase
+                 * postpone to next interval */
+                if (time_until_tx < LWMAC_WR_PREPARATION_US) {
+                    time_until_tx += LWMAC_WAKEUP_INTERVAL_US;
+                }
+
+                time_until_tx -= LWMAC_WR_PREPARATION_US;
+                lwmac_set_timeout(&lwmac, TIMEOUT_WAIT_FOR_DEST_WAKEUP, time_until_tx);
+
+                /* Register neighbour to be the next */
+                lwmac.tx.current_neighbour = neighbour;
+
+                /* Stop dutycycling, we're preparing to send. This prevents the
+                 * timeout arriving late, so that the destination phase would
+                 * be missed. */
```

Comment:
> So, in the standby mode, xtimer (timeout module) is still working, but may be dull (the timer interrupt and related reaction/execution may come late due to low frequency of CPU?) can I understand it like that? :-)

This actually depends on how you have configured the timer hardware, but for most cases xtimer will just stop working, i.e. the system time will not be incremented and timer alarms won't be generated.

> Will setting the lmp mode, i.e., `lpm_prevent_sleep &= ~(LWMAC_LPM_MASK)`, put the CPU into a low frequency in which the thread of LwMAC is still running with a low frequency, or, in the lmp mode, the cpu totally stop (only RTT is running)? This also depends on the level of lmp, right? 

Almost the same as above: usually it will stop the cpu, but may depend on the implementation. But the idea of "sleep" is to stop I'd say.


Usually, the xtimer runs from the same clock source as the cpu. So this clock source is stopped and both xtimer and cpu stops. The RTT however runs from another clock, usually the 32kHz low power clock and keeps running.